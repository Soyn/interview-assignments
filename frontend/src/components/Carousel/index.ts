export {
  Caroussel
} from './Carousel';

export type { ICarouselProps as CarouselProps } from './Carousel';