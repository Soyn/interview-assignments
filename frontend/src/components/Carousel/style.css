.carousel-card {
  transition: transform 0.6s ease-in-out;
}

.active,
.previous,
.next {
  display: block;
}

.next,
.previous {
  position: absolute;
  top: 0;
}

.active {
  transform: translate(0%,0%);
}

.previous {
  transform: translate(-100%,0%);
}
.next {
  transform: translate(100%,0%);
}

.indicators {
  bottom: 3em;
  left: 50%;
  transform: translateX(-50%);
}

.indicators .indicator-inner {
  transition: width;
}