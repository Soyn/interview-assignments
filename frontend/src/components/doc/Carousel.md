# Carousel
## Example
### With Controls
### With Indicators
### Cycle
### Options