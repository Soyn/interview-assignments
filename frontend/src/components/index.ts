export {
  Caroussel
} from './Carousel';
export type { CarouselProps } from './Carousel';
