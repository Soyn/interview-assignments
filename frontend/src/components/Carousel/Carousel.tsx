import React, {
  useState,
  useRef,
  useEffect,
  KeyboardEvent,
  MutableRefObject,
  MouseEvent,
} from "react";
import classNames from "classnames";

import "./style.css";

export interface ICarouselProps {
  interval?: number; // ms
  items: JSX.Element[];
  customCSS?: string;
  keyboard?: boolean;
  indicators?: boolean;
  control?: boolean;
  pasued?: string | boolean;
  cycle?: boolean;
}

const EmptyPage = () => (
  <div className="container text-center">Opps! Nothing!</div>
);

export function Carousel(props: ICarouselProps) {
  const {
    customCSS,
    items,
    interval,
    keyboard,
    indicators,
    control,
    cycle,
    pasued,
  } = props;
  const [pageIndex, setPageIndex] = useState(0);
  const inputRef = useRef<HTMLInputElement>(null);
  let timer: MutableRefObject<number | null> = useRef<number>(null);
  let isPausedCyCle: MutableRefObject<boolean> = useRef<boolean>(false);
  const toNext = () => {
    setPageIndex(pageIndex + 1);
  };
  const toPrevious = () => {
    let idx = pageIndex - 1;
    if (idx < 0) {
      idx = items.length - 1;
    }
    setPageIndex(idx);
  };
  const clearCycle = () => {
    if (timer.current !== null) {
      clearTimeout(timer.current);
    }
  };
  const pauseCycle = () => {
    if (pasued === "hover") {
      isPausedCyCle.current = true;
      clearCycle();
    }
  };
  const startCycle = () => {
    isPausedCyCle.current = false;
    if (cycle && pasued === "hover") {
      clearCycle();
      toNext();
    }
  };
  const onKeyUp = (e: KeyboardEvent) => {
    e.preventDefault();
    if (keyboard) {
      const { key } = e;
      const isHitKeyPrompt = key === "ArrowRight" || key === "ArrowLeft";
      if (isHitKeyPrompt) {
        clearCycle();
        switch (key) {
          case "ArrowLeft": {
            toPrevious();
            break;
          }
          case "ArrowRight": {
            toNext();
            break;
          }
        }
      }
    }
  };
  const getFocus = (e: MouseEvent) => {
    e.stopPropagation();
    if (inputRef?.current) {
      inputRef.current.focus();
    }
  };
  const nav2Prev = () => {
    clearCycle();
    toPrevious();
  };
  const nav2Next = () => {
    clearCycle();
    toNext();
  };
  // schedule next swip
  useEffect(() => {
    if (cycle && !isPausedCyCle.current) {
      timer.current = window.setTimeout(() => {
        toNext();
      }, interval);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pageIndex]);
  useEffect(() => {
    if (inputRef) {
      inputRef.current?.focus();
    }
  }, []);
  if (items.length === 0) return <EmptyPage />;
  return (
    <div
      className={classNames(
        "w-full h-full relative overflow-hidden",
        customCSS || ""
      )}
      onClick={getFocus}
      onMouseEnter={pauseCycle}
      onMouseLeave={startCycle}
    >
      {keyboard ? (
        <input
          className="keyboard-dector absolute focus:no-underline focus:outline-none"
          onKeyUp={onKeyUp}
          ref={inputRef}
        />
      ) : null}
      <div className="w-full h-full relative overflow-hidden">
        {items.map((e, i) => {
          const activeCardIdx = pageIndex % items.length;
          let animation =
            activeCardIdx === i
              ? "active"
              : i < activeCardIdx
              ? "previous"
              : "next";
          return (
            <div
              className={classNames(
                "w-full h-full carousel-card hidden",
                animation
              )}
              key={i}
            >
              {e}
            </div>
          );
        })}
      </div>
      {indicators ? (
        <div className="indicators flex flex-row items-center absolute p-1">
          {items.map((_, i) => {
            const duration = cycle ? `${interval}ms` : "0";
            const style = {
              transitionDuration: duration,
            };
            return (
              <div className="m-1 w-14 h-0.5 bg-gray-400" key={i}>
                <div
                  className={classNames(
                    "h-full indicator-inner",
                    i === pageIndex % items.length ? "bg-gray-50 w-full" : "w-0"
                  )}
                  style={style}
                />
              </div>
            );
          })}
        </div>
      ) : null}
      {control ? (
        <>
          <button
            className="control-prev
            absolute w-1/5 h-full
            flex justify-start items-center
            text-gray-400
            opacity-0 hover:opacity-100 transition-opacity duration-1000"
            data-testid="prev"
            onClick={nav2Prev}
            role="link"
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="h-12 w-12"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M15 19l-7-7 7-7"
              />
            </svg>
          </button>
          <button
            className="control-next
            absolute w-1/5 h-full 
            flex justify-end items-center text-gray-400
            opacity-0 hover:opacity-100 transition-opacity duration-1000"
            data-testid="next"
            onClick={nav2Next}
            role="link"
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="h-12 w-12"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M9 5l7 7-7 7"
              />
            </svg>
          </button>
        </>
      ) : null}
    </div>
  );
}

Carousel.defaultProps = {
  interval: 1000,
};
