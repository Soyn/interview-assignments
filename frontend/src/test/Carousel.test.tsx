import React from "react";
import { create, act } from "react-test-renderer";
import { cleanup, fireEvent, render, waitFor, prettyDOM } from "@testing-library/react";
import { Caroussel } from "../components/Carousel";
import { getConstantValue } from "typescript";

afterEach(() => {
  cleanup();
});

test("render empty page, if no item was passed", () => {
  const component = create(<Caroussel items={[]} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("render right tree structure when passed items", () => {
  const getSlides = () => {
    const descriptions = [];
    for (let i = 0; i < 3; i += 1) {
      descriptions.push({
        text: `test-${i}`,
        bgColor: "red",
      });
    }
    return descriptions.map(({ text, bgColor }, i) => (
      <div
        key={i}
        style={{
          backgroundColor: bgColor,
        }}
      >
        {text}
      </div>
    ));
  };
  let root;
  act(() => {
    root = create(<Caroussel items={getSlides()} />);
  });
  expect(root.toJSON()).toMatchSnapshot();

  act(() => {
    root.update(<Caroussel items={getSlides()} indicators />);
  });
  expect(root.toJSON()).toMatchSnapshot();

  act(() => {
    root.update(<Caroussel items={getSlides()} control />);
  });
  expect(root.toJSON()).toMatchSnapshot();

  act(() => {
    root.update(<Caroussel items={getSlides()} control indicators />);
  });
  expect(root.toJSON()).toMatchSnapshot();
});

test("do interaction when user slide the page manually", () => {
  const getSlides = () => {
    const descriptions = [];
    for (let i = 1; i <= 3; i += 1) {
      descriptions.push({
        text: `test-${i}`,
        bgColor: "red",
      });
    }
    return descriptions.map(({ text, bgColor }, i) => (
      <div
        key={i}
        style={{
          backgroundColor: bgColor,
        }}
      >
        {text}
      </div>
    ));
  };
  const { queryAllByRole, queryByTestId, getByText } = render(
    <Caroussel items={getSlides()} control />
  );
  expect(queryAllByRole("link").length).toBe(2);

  const prev = queryByTestId("prev");
  fireEvent.click(prev);
  expect(
    getByText("test-3")?.parentElement?.classList.contains("active")
  ).toBeTruthy();

  const next = queryByTestId("next");
  fireEvent.click(next);
  expect(
    getByText("test-1")?.parentElement?.classList.contains("active")
  ).toBeTruthy();

  fireEvent.click(next);
  fireEvent.click(next);
  fireEvent.click(next);
  expect(
    getByText("test-1")?.parentElement?.classList.contains("active")
  ).toBeTruthy();
});

test("Cycle the slides", async () => {
  const getSlides = () => {
    const descriptions = [];
    for (let i = 1; i <= 3; i += 1) {
      descriptions.push({
        text: `test-${i}`,
        bgColor: "red",
      });
    }
    return descriptions.map(({ text, bgColor }, i) => (
      <div
        key={i}
        style={{
          backgroundColor: bgColor,
        }}
      >
        {text}
      </div>
    ));
  };
  const { container, getByText } = render(
    <Caroussel items={getSlides()} cycle interval={500} />
  );
  const test1 = getByText('test-1').parentElement;
  const test2 = getByText('test-2').parentElement;
  const test3 = getByText('test-3').parentElement;

  await waitFor(() => {
    expect(test1?.classList.contains('active')).toBeTruthy();
    expect(test2?.classList.contains('next')).toBeTruthy();
    expect(test3?.classList.contains('next')).toBeTruthy();
  }, { timeout: 400 })

  await waitFor(() => {

    expect(test1?.classList.contains('previous')).toBeTruthy();
    expect(test2?.classList.contains('active')).toBeTruthy();
    expect(test3?.classList.contains('next')).toBeTruthy();
  }, {
    timeout: 600,
  });

  await waitFor(() => {

    expect(test1?.classList.contains('previous')).toBeTruthy();
    expect(test2?.classList.contains('previous')).toBeTruthy();
    expect(test3?.classList.contains('active')).toBeTruthy();
  }, {
    timeout: 1200,
  });

  await waitFor(() => {
    expect(test1?.classList.contains('active')).toBeTruthy();
    expect(test2?.classList.contains('next')).toBeTruthy();
    expect(test3?.classList.contains('next')).toBeTruthy();
  }, {
    timeout: 1500,
  });
});
