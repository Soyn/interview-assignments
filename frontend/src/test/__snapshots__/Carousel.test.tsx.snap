// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render empty page, if no item was passed 1`] = `
<div
  className="container text-center"
>
  Opps! Nothing!
</div>
`;

exports[`render right tree structure when passed items 1`] = `
<div
  className="w-full h-full relative overflow-hidden"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="w-full h-full relative overflow-hidden"
  >
    <div
      className="w-full h-full carousel-card hidden active"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-0
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-1
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-2
      </div>
    </div>
  </div>
</div>
`;

exports[`render right tree structure when passed items 2`] = `
<div
  className="w-full h-full relative overflow-hidden"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="w-full h-full relative overflow-hidden"
  >
    <div
      className="w-full h-full carousel-card hidden active"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-0
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-1
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-2
      </div>
    </div>
  </div>
  <div
    className="indicators flex flex-row items-center absolute p-1"
  >
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner bg-red-300 w-full"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner w-0"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner w-0"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`render right tree structure when passed items 3`] = `
<div
  className="w-full h-full relative overflow-hidden"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="w-full h-full relative overflow-hidden"
  >
    <div
      className="w-full h-full carousel-card hidden active"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-0
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-1
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-2
      </div>
    </div>
  </div>
  <a
    className="control-prev absolute w-1/5 h-full"
    data-testid="prev"
    onClick={[Function]}
    role="link"
  >
    <span />
  </a>
  <a
    className="control-next absolute w-1/5 h-full"
    data-testid="next"
    onClick={[Function]}
    role="link"
  >
    <span />
  </a>
</div>
`;

exports[`render right tree structure when passed items 4`] = `
<div
  className="w-full h-full relative overflow-hidden"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="w-full h-full relative overflow-hidden"
  >
    <div
      className="w-full h-full carousel-card hidden active"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-0
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-1
      </div>
    </div>
    <div
      className="w-full h-full carousel-card hidden next"
    >
      <div
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        test-2
      </div>
    </div>
  </div>
  <div
    className="indicators flex flex-row items-center absolute p-1"
  >
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner bg-red-300 w-full"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner w-0"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
    <div
      className="m-1 w-12 h-2 bg-gray-400"
    >
      <div
        className="h-full indicator-inner w-0"
        style={
          Object {
            "transitionDuration": "0",
          }
        }
      />
    </div>
  </div>
  <a
    className="control-prev absolute w-1/5 h-full"
    data-testid="prev"
    onClick={[Function]}
    role="link"
  >
    <span />
  </a>
  <a
    className="control-next absolute w-1/5 h-full"
    data-testid="next"
    onClick={[Function]}
    role="link"
  >
    <span />
  </a>
</div>
`;
